docs/cloud-integration.md

```diff
@@ -0,0 +1,512 @@
+using a URL to the data in methods like `SparkContext.textFile()` to read data, 
+`saveAsTextFile()` to write it back.
+
+
+Because object stores are viewed by Spark as filesystems, object stores can
+be used as the source or destination of any spark work —be it batch, SQL, DataFrame,
+Streaming or something else.
+
+The steps to do so are as follows
+
+1. Use the full URI to refer to a bucket, including the prefix for the client-side library
+to use. Example: `s3a://landsat-pds/scene_list.gz`
+1. Have the Spark context configured with the authentication details of the object store.
+In a YARN cluster, this may also be done in the `core-site.xml` file.
+
+
+## <a name="output"></a>Object Stores as a substitute for HDFS
+
+As the examples show, you can write data to object stores. However, that does not mean
+That they can be used as replacements for a cluster-wide filesystem.
+
+The full details are covered in [Cloud Object Stores are Not Real Filesystems](#cloud_stores_are_not_filesystems).
+
+The brief summary is:
+
+| Object Store Connector      |  Replace HDFS? |
+|-----------------------------|--------------------|
+| `s3a://` `s3n://`  from the ASF   | No  |
+| Amazon EMR `s3://`          | Yes |
+| Microsoft Azure `wasb://`   | Yes |
+| OpenStack `swift://`        | No  |
+
+It is possible to use any of the object stores as a destination of work, i.e. use
+`saveAsTextFile()` or `save()` to save data there, but the commit process may be slow
+and, unreliable in the presence of failures.
+
+It is faster and safer to use the cluster filesystem as the destination of Spark jobs,
+using that data as the data for follow-on work. The final results can
+be persisted in to the object store using `distcp`.
+
+#### <a name="checkpointing"></a>Spark Streaming and object stores
+
+Spark Streaming can monitor files added to object stores, by
+creating a `FileInputDStream` DStream monitoring a path under a bucket through
+`StreamingContext.textFileStream()`.
+
+
+1. The time to scan for new files is proportional to the number of files
+under the path —not the number of *new* files, and that it can become a slow operation.
+The size of the window needs to be set to handle this.
+
+1. Files only appear in an object store once they are completely written; there
+is no need for a worklow of write-then-rename to ensure that files aren't picked up
+while they are still being written. Applications can write straight to the monitored directory.
+
+1. Streams should only be checkpointed to an object store considered compatible with
+HDFS. Otherwise the checkpointing will be slow and potentially unreliable.
+
+### Recommended settings for writing to object stores
+
+Here are the settings to use when writing to object stores. This uses the "version 2" algorithm
+for committing files —which does less renaming than the v1 algorithm. Speculative execution is
+disabled to avoid multiple writers corrupting the output.
```

Comment:
See [SPARK-4879](https://issues.apache.org/jira/browse/SPARK-4879) for this surfacing in Spark & S3, albeit before Josh added the coordination layer in Spark. The underlying committer was using that O(1) rename as an implicit co-ordinator amongst workers, which doesn't work when its a non-atomic O(file *data) kind of call.

Thinking about it, I could just be over-cautious based on my fear of the current s3a/FileOutputCommitter knowledge; other object stores may behave better (Azure has leases to manage some of this), and Josh's protocol should make the problem go away on all but the direct committer, which is disabled when speculate=true.

so, 3 options

1. Don't mention it at all
2. Say "you should do this"
3. Say "speculation may not work, consult your cloud storage provider/connector docs"?

option 3 means that when you use the current netflix staging committer, my s3guard derivative, etc, all is well, same for anything that others provide. And I'll be doing my best to test all of this to see if I can create a problem